docs/route/middleware.md

```diff
@@ -0,0 +1,107 @@
+      let user = context.get(userContext);
+      let data = await getProfile(user);
+      return json(data);
+    }}
+  />
+</Route>;
+```
+
+## Arguments
+
+Middleware receives the same arguments as a `loader`/`action` (`request` and `params`) as well as an additional `context` parameter. `context` behaves a bit like [React Context][react-context] in that you create a context which is the strongly-typed to the value you provide.
+
+```tsx
+let userContext = createMiddlewareContext<User>();
+// context.get(userContext)       => Return type is User
+// context.set(userContext, user) => Requires `user` be of type User
+```
+
+When middleware is enabled, this `context` object is also made available to actions and loaders to retrieve values set by middlewares.
+
+## Logic Flow
+
+Middleware is designed to solve 4 separate use-cases with a single API to keep the API surface compact:
+
+- I want to run some logic before a loader
+- I want to run some logic after a loader
+- I want to run some logic before an action
+- I want to run some logic after an action
+
+To support this we adopted an API inspired by the middleware implementation in [Fresh][fresh-middleware] where the function gives you control over the invocation of child logic, thus allowing you to run logic both before and after the child logic. You can differentiate between loaders and actions based on the `request.method`.
+
+```tsx
+async function middleware({ request, context }) {
+  // Run me top-down before action/loaders run
+  // Provide values to descendant middleware + action/loaders
+  context.set(userContext, getUser(request));
+
+  // Run descendant middlewares sequentially, followed by the action/loaders
+  let response = await context.next();
+
+  // Run me bottom-up after action/loaders run
+  response.headers.set("X-Custom", "Stuff");
+
+  // Return the response to ancestor middlewares
+  return response;
+}
+```
+
+Because middleware has access to the incoming `Request` _and also_ has the ability to mutate the outgoing `Response`, it's important to note that middlewares are executed _per-unique Request/Response combination_.
+
+In client-side React Router applications, this means that nested routes will execute middlewares _for each loader_ because each loader returns a unique `Response` that could be altered independently by the middleware.
+
+When navigating to `/a/b`, the following represents the parallel data loading chains:
+
+```
+a middleware -> a loader -> a middleware
+a middleware -> b middleware -> b loader -> b middleware -> a middleware
+```
+
+So you should be aware that while middleware will reduce some code duplication across your actions/loaders, you may need to leverage a mechanism to dedup external API calls made from within a middleware.
+
+[pickingarouter]: ../routers/picking-a-router
+[react-context]: https://reactjs.org/docs/context.html
+[fresh-middleware]: https://fresh.deno.dev/docs/concepts/middleware
```

Comment:
What you would middleware composition look like with this new API?
For example, what If I needed to apply 3-4 middleware to my request, which is not an uncommon thing if you're coming from Express or Fastify or in general a production nodejs server 😄 

https://github.com/data-eden/data-eden/tree/main/packages/network
![CleanShot 2023-02-28 at 04 01 19@2x](https://user-images.githubusercontent.com/6743796/221849142-62b5c2a2-7431-4089-8b22-d58cc67c07e1.png)